stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -943,6 +943,35 @@ void swift::_swift_taskGroup_detachChild(TaskGroup *group,
   });
 }
 
+/// Cancel all the child tasks that belong to `group`.
+///
+/// The caller must guarantee that this is called while holding the owning
+/// task's status record lock.
+void swift::_swift_taskGroup_cancelAllChildren(TaskGroup *group) {
+  // Because only the owning task of the task group can modify the
+  // child list of a task group status record, and it can only do so
+  // while holding the owning task's status record lock, we do not need
+  // any additional synchronization within this function.
+  for (auto childTask : group->getTaskRecord()->children())
+    swift_task_cancel(childTask);
+}
+
+/// Cancel all the child tasks that belong to `group`.
+///
+/// The caller must guarantee that this is called from the owning task.
+void swift::_swift_taskGroup_cancelAllChildren_unlocked(TaskGroup *group,
+                                                        AsyncTask *owningTask) {
+  // Early out. If there are no children, there's nothing to do. We can safely
+  // check this without locking, since this can only be concurrently mutated
+  // from a child task. If there are no children then no more can be added.
+  if (!group->getTaskRecord()->getFirstChild())
+    return;
+
+  withStatusRecordLock(owningTask, [&group](ActiveTaskStatus status) {
+    _swift_taskGroup_cancelAllChildren(group);
+  });
+}
```

Comment:
Hmm, we use a different naming scheme with the status records in `TaskPrivate.h` so the locked ones are without suffix and the ones "assume to call with a lock" are Locked, but I guess `_unlocked` is fine, since we want to keep `cancelAllChildren` I guess.

Sounds okey, just got me thinking about naming

Reply:
Bit of a tangent, but the ObjC runtime has a crazy scheme where debug builds wrap all locks with a debug lock type which tracks which locks are owned by the current thread, and then functions can easily assert that a given lock is held or not held when it expects. That sort of thing could be helpful her. (Of course, you still have to realize that you require that lock to be held, and write the assert.)

Reply:
I like that, we could sprinkle them over some methods which access records etc. Could be something nice to follow up on